### ISSUE TITLE
Generic typing for addPromise

### ISSUE THREAD

**Issue by IlyaSemenov** [doo-gl/concurrent-promise-queue#1]
I believe the generic typing in the library is incorrect. Currently, the return type is bound at instance level:

https://github.com/doo-gl/concurrent-promise-queue/blob/7beabd8bb07493c29a8edbf27a90e3fb59345361/src/index.ts#L37

then `addPromise` is supposed to only take promises of the given type:

https://github.com/doo-gl/concurrent-promise-queue/blob/7beabd8bb07493c29a8edbf27a90e3fb59345361/src/index.ts#L72

There is no point in limiting the whole queue to a single type. It's very natural for the queue to be handling promises of different types. For example, if it's a remote API access throttle queue, every API request will return its own type, which is currently not supported.

Consider this example:

```ts
const q = new ConcurrentPromiseQueue({})
const value1 = await q.addPromise(async () => 123)
const value2 = await q.addPromise(async () => "foo")
```

Both `value1` and `value2` are now typed as `unknown` which is a bad DX.

In my thinking,
1) there is no point of having `T` at the queue instance level.
2) `addPromise` should be a generic of its own.

This is how I currently extend the class to be suitable for real use:

```ts
class TypedConcurrentPromiseQueue extends ConcurrentPromiseQueue<unknown> {
  override addPromise<T>(promiseSupplier: PromiseSupplier<T>): Promise<T> {
    return super.addPromise(promiseSupplier) as Promise<T>
  }
}

const q = new TypedConcurrentPromiseQueue({})
const value1 = await q.addPromise(async () => 123)
const value2 = await q.addPromise(async () => "foo")
```

Now `value1` is typed as a number and `value2` is typed as a string, as expected.

I propose to update the original typings accordingly.

---

Note that in your typings, `addPromise` returns `Promise<T | null>` which is also apparently invalid. The only way the promises resolve is through `promiseExecutionListener`, and there is no case when `promiseExecutionListener` will successfully resolve with `null` and not the original promise return type `T`.

**Comment by IlyaSemenov** [comment]
I presume this project is abandoned? Is there a maintained alternative you could possibly recommend instead?

**Comment by IlyaSemenov** [comment]
For anyone else interested, `p-queue` seems to be working similarly:

```ts
import Queue from "p-queue"

const queue = new Queue({
  concurrency: 1,
  interval: 1500,
  intervalCap: 1,
})

// typed as number
const value1 = await queue.add(async () => 123, { throwOnTimeout: true })

// typed as string
const value2 = await queue.add(async () => "foo", { throwOnTimeout: true })
```

It's ugly that one has to use `throwOnTimeout` all the time, though.

**Comment by doo-gl** [comment]
hi, apologies, this was a util i generated for a project that i land up reusing, I will take a look at your suggestions here and the other issue and try create an improved version

**Comment by doo-gl** [comment]
Published version v1.1.0 which adds the generic typing described, closing this as a result.
